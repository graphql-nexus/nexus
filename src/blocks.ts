export { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
export { InterfaceDefinitionBlock } from './definitions/interfaceType'
export { ObjectDefinitionBlock } from './definitions/objectType'
export { UnionDefinitionBlock } from './definitions/unionType'
