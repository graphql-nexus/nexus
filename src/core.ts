// The "core" is used as a namespace to re-export everything,
// For anyone who wants to use the internals
export * from './builder'
export * from './makeSchema'
export * from './definitions/args'
export * from './definitions/decorateType'
export * from './definitions/definitionBlocks'
export * from './definitions/enumType'
export * from './definitions/extendInputType'
export * from './definitions/extendType'
export * from './definitions/inputObjectType'
export * from './definitions/interfaceType'
export * from './definitions/list'
export * from './definitions/mutationField'
export * from './definitions/mutationType'
export * from './definitions/nonNull'
export * from './definitions/nullable'
export * from './definitions/objectType'
export * from './definitions/queryField'
export * from './definitions/queryType'
export * from './definitions/scalarType'
export * from './definitions/subscriptionField'
export * from './definitions/subscriptionType'
export * from './definitions/nexusMeta'
export * from './definitions/unionType'
export * from './definitions/wrapping'
export * from './definitions/_types'
export * from './dynamicMethod'
export * from './plugin'
export * from './plugins'
export * from './rebuildType'
export * from './sdlConverter'
export * from './typegenAutoConfig'
export * from './typegenFormatPrettier'
export * from './typegenMetadata'
export * from './typegenPrinter'
export * from './typegenTypeHelpers'
export * from './typegenUtils'
export * from './utils'
