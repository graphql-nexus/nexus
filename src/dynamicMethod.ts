import { PluginBuilderLens } from './builder'
import { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'

export type OutputFactoryConfig<T> = {
  stage: 'walk' | 'build'
  args: any[]
  builder: PluginBuilderLens
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export type InputFactoryConfig<T> = {
  args: any[]
  builder: PluginBuilderLens
  typeDef: InputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface BaseExtensionConfig<T extends string> {
  /**
   * The name of the "extension", the field made
   * available on the builders
   */
  name: T
  /**
   * The full type definition for the options, including generic
   * signature for the type
   */
  typeDefinition?: string
}

export interface DynamicOutputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: OutputFactoryConfig<T>): any
}

export interface DynamicInputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: InputFactoryConfig<T>): any
}

export class DynamicInputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicInputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicInputMethodDef, NexusTypes.DynamicInput)

export class DynamicOutputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputMethodDef, NexusTypes.DynamicOutputMethod)

/**
 * Defines a new property on the object definition block
 * for an output type, taking arbitrary input to define
 * additional types. See the connectionPlugin:
 *
 * t.connectionField('posts', {
 *   nullable: true,
 *   totalCount(root, args, ctx, info) {
 *     return ctx.user.getTotalPostCount(root.id, args)
 *   },
 *   nodes(root, args, ctx, info) {
 *     return ctx.user.getPosts(root.id, args)
 *   }
 * })
 */
export function dynamicOutputMethod<T extends string>(config: DynamicOutputMethodConfig<T>) {
  return new DynamicOutputMethodDef(config.name, config)
}

/**
 * Same as the outputFieldExtension, but for fields that
 * should be added on as input types.
 */
export function dynamicInputMethod<T extends string>(config: DynamicInputMethodConfig<T>) {
  return new DynamicInputMethodDef(config.name, config)
}
