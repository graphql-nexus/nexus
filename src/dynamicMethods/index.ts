export * from './collection'
export * from './relayConnection'
