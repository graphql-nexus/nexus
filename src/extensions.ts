import { GraphQLNamedType } from 'graphql'
import { DynamicFieldDefs, SchemaConfig } from './builder'
import { NexusOutputFieldConfig } from './definitions/definitionBlocks'
import { NexusInputObjectTypeConfig } from './definitions/inputObjectType'
import { NexusInterfaceTypeConfig } from './definitions/interfaceType'
import { NexusObjectTypeConfig } from './definitions/objectType'
import { RootTypings } from './definitions/_types'

export type NexusGraphQLNamedType = GraphQLNamedType & {
  extensions?: {
    nexus?: {
      config: any
    }
  }
}

export type NexusTypeExtensions = NexusObjectTypeExtension | NexusInterfaceTypeExtension

/**
 * Container object living on `fieldDefinition.extensions.nexus`
 */
export class NexusFieldExtension<TypeName extends string = any, FieldName extends string = any> {
  readonly config: Omit<NexusOutputFieldConfig<TypeName, FieldName>, 'resolve'>
  /**
   * Whether the user has provided a custom "resolve" function,
   * or whether we're using GraphQL's defaultResolver
   */
  readonly hasDefinedResolver: boolean
  constructor(config: NexusOutputFieldConfig<TypeName, FieldName>) {
    const { resolve, ...rest } = config
    this.config = rest
    this.hasDefinedResolver = Boolean(resolve)
  }
}

/**
 * Container object living on `inputObjectType.extensions.nexus`
 */
export class NexusInputObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInputObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInputObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `objectType.extensions.nexus`
 */
export class NexusObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `interfaceType.extensions.nexus`
 */
export class NexusInterfaceTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInterfaceTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInterfaceTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

export interface NexusSchemaExtensionConfig extends Omit<SchemaConfig, 'types'> {
  dynamicFields: DynamicFieldDefs
  rootTypings: RootTypings
}

/**
 * Container object living on `schema.extensions.nexus`. Keeps track
 * of metadata from the builder so we can use it when we
 */
export class NexusSchemaExtension {
  constructor(readonly config: NexusSchemaExtensionConfig) {}
}
