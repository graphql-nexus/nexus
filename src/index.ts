import * as blocks from './blocks'
import * as core from './core'

// All of the Public API definitions
export { makeSchema } from './makeSchema'
export { arg, booleanArg, floatArg, idArg, intArg, stringArg } from './definitions/args'
export { decorateType } from './definitions/decorateType'
export { enumType } from './definitions/enumType'
export { extendInputType } from './definitions/extendInputType'
export { extendType } from './definitions/extendType'
export { inputObjectType } from './definitions/inputObjectType'
export { interfaceType } from './definitions/interfaceType'
export { list } from './definitions/list'
export { mutationField } from './definitions/mutationField'
export { mutationType } from './definitions/mutationType'
export { nonNull } from './definitions/nonNull'
export { nullable } from './definitions/nullable'
export { objectType } from './definitions/objectType'
export { queryField } from './definitions/queryField'
export { queryType } from './definitions/queryType'
export { asNexusMethod, scalarType } from './definitions/scalarType'
export { subscriptionField } from './definitions/subscriptionField'
export { subscriptionType } from './definitions/subscriptionType'
export { unionType } from './definitions/unionType'
export { dynamicInputMethod, dynamicOutputMethod } from './dynamicMethod'
export { dynamicOutputProperty } from './dynamicProperty'
export { createPlugin, plugin, PluginBuilderLens, PluginConfig } from './plugin'
export * from './plugins'
export { convertSDL } from './sdlConverter'
export { AllInputTypes, AllOutputTypes, FieldResolver, FieldType } from './typegenTypeHelpers'
export { groupTypes } from './utils'
export { blocks, core }
