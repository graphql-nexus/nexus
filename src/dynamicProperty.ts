import { SchemaBuilder } from './builder'
import { OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'
import { BaseExtensionConfig } from './dynamicMethod'

export type OutputPropertyFactoryConfig<T> = {
  stage: 'walk' | 'build'
  builder: SchemaBuilder
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface DynamicOutputPropertyConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the property is accessed (as a getter)
   */
  factory(config: OutputPropertyFactoryConfig<T>): any
}

export class DynamicOutputPropertyDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputPropertyConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputPropertyDef, NexusTypes.DynamicOutputProperty)

/**
 * Defines a new property on the object definition block
 * for an output type, making it possible to build custom DSL's
 * on top of Nexus, e.g. in nexus-prisma
 *
 * t.model.posts()
 */
export function dynamicOutputProperty<T extends string>(config: DynamicOutputPropertyConfig<T>) {
  return new DynamicOutputPropertyDef(config.name, config)
}
