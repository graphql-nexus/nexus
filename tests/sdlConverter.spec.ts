import { convertSDL, core } from "../src";
import { EXAMPLE_SDL } from "./_sdl";

const { SDLConverter } = core;

describe("SDLConverter", () => {
  test("printObjectTypes", () => {
    expect(new SDLConverter(EXAMPLE_SDL).printObjectTypes())
      .toMatchInlineSnapshot(`
"export const Mutation = objectType({
  name: \\"Mutation\\",
  definition(t) {
    t.string(\\"someList\\", {
      list: [false],
      args: {
        items: stringArg({
          list: [false],
          required: true
        }),
      },
    })
    t.field(\\"createPost\\", {
      type: Post,
      args: {
        input: arg({
          type: CreatePostInput,
          required: true
        }),
      },
    })
    t.field(\\"registerClick\\", {
      type: Query,
      args: {
        uuid: uuidArg(),
      },
    })
  }
})
export const Post = objectType({
  name: \\"Post\\",
  definition(t) {
    t.implements(Node)
    t.uuid(\\"uuid\\")
    t.field(\\"author\\", { type: User })
    t.float(\\"geo\\", { list: [true, true] })
    t.float(\\"messyGeo\\", {
      list: [true, false],
      nullable: true,
    })
  }
})
export const Query = objectType({
  name: \\"Query\\",
  definition(t) {
    t.field(\\"user\\", { type: User })
    t.list.field(\\"posts\\", {
      type: Post,
      args: {
        filters: arg({
          type: PostFilters,
          required: true
        }),
      },
    })
    t.field(\\"unionField\\", { type: ExampleUnion })
  }
})
export const User = objectType({
  name: \\"User\\",
  definition(t) {
    t.implements(Node)
    t.string(\\"name\\", {
      description: \\"This is a description of a name\\",
      args: {
        prefix: stringArg({ description: \\"And a description of an arg\\" }),
      },
    })
    t.string(\\"email\\")
    t.string(\\"phone\\", { nullable: true })
    t.list.field(\\"posts\\", {
      type: Post,
      args: {
        filters: arg({ type: PostFilters }),
      },
    })
    t.field(\\"outEnum\\", {
      type: SomeEnum,
      nullable: true,
    })
  }
})"
`);
  });

  test("printEnumTypes", () => {
    expect(new SDLConverter(EXAMPLE_SDL).printEnumTypes())
      .toMatchInlineSnapshot(`
"export const OrderEnum = enumType({
  name: \\"OrderEnum\\",
  members: [\\"ASC\\",\\"DESC\\"],
});
export const SomeEnum = enumType({
  name: \\"SomeEnum\\",
  members: [\\"A\\",\\"B\\"],
});"
`);
  });

  test("printScalarTypes", () => {
    expect(new SDLConverter(EXAMPLE_SDL).printScalarTypes())
      .toMatchInlineSnapshot(`
"export const UUID = scalarType({
  name: \\"UUID\\",
  asNexusMethod: \\"uuid\\",
  serialize() { /* Todo */ },
  parseValue() { /* Todo */ },
  parseLiteral() { /* Todo */ }
});"
`);
  });

  test("printInterfaceTypes", () => {
    expect(new SDLConverter(EXAMPLE_SDL).printInterfaceTypes())
      .toMatchInlineSnapshot(`
"export const Node = interfaceType({
  name: \\"Node\\",
  description: \\"This is a description of a Node\\",
  definition(t) {
    t.id(\\"id\\")
    t.resolveType(() => null)
  }
});"
`);
  });
});

test("convertSDL", () => {
  expect(convertSDL(EXAMPLE_SDL)).toMatchInlineSnapshot(`
"import { objectType, stringArg, arg, uuidArg, interfaceType, inputObjectType, unionType, enumType, scalarType } from 'nexus';

export const Mutation = objectType({
  name: \\"Mutation\\",
  definition(t) {
    t.string(\\"someList\\", {
      list: [false],
      args: {
        items: stringArg({
          list: [false],
          required: true
        }),
      },
    })
    t.field(\\"createPost\\", {
      type: Post,
      args: {
        input: arg({
          type: CreatePostInput,
          required: true
        }),
      },
    })
    t.field(\\"registerClick\\", {
      type: Query,
      args: {
        uuid: uuidArg(),
      },
    })
  }
})
export const Post = objectType({
  name: \\"Post\\",
  definition(t) {
    t.implements(Node)
    t.uuid(\\"uuid\\")
    t.field(\\"author\\", { type: User })
    t.float(\\"geo\\", { list: [true, true] })
    t.float(\\"messyGeo\\", {
      list: [true, false],
      nullable: true,
    })
  }
})
export const Query = objectType({
  name: \\"Query\\",
  definition(t) {
    t.field(\\"user\\", { type: User })
    t.list.field(\\"posts\\", {
      type: Post,
      args: {
        filters: arg({
          type: PostFilters,
          required: true
        }),
      },
    })
    t.field(\\"unionField\\", { type: ExampleUnion })
  }
})
export const User = objectType({
  name: \\"User\\",
  definition(t) {
    t.implements(Node)
    t.string(\\"name\\", {
      description: \\"This is a description of a name\\",
      args: {
        prefix: stringArg({ description: \\"And a description of an arg\\" }),
      },
    })
    t.string(\\"email\\")
    t.string(\\"phone\\", { nullable: true })
    t.list.field(\\"posts\\", {
      type: Post,
      args: {
        filters: arg({ type: PostFilters }),
      },
    })
    t.field(\\"outEnum\\", {
      type: SomeEnum,
      nullable: true,
    })
  }
})

export const Node = interfaceType({
  name: \\"Node\\",
  description: \\"This is a description of a Node\\",
  definition(t) {
    t.id(\\"id\\")
    t.resolveType(() => null)
  }
});

export const CreatePostInput = inputObjectType({
  name: \\"CreatePostInput\\",
  definition(t) {
    t.string(\\"name\\", { required: true })
    t.id(\\"author\\", { required: true })
    t.float(\\"geo\\", {
      list: [false, true],
      required: true,
    })
  }
});
export const PostFilters = inputObjectType({
  name: \\"PostFilters\\",
  definition(t) {
    t.field(\\"order\\", {
      type: OrderEnum,
      required: true,
    })
    t.string(\\"search\\")
  }
});

export const ExampleUnion = unionType({
  name: \\"ExampleUnion\\",
  definition(t) {
    t.members(Post, User)
  }
});

export const OrderEnum = enumType({
  name: \\"OrderEnum\\",
  members: [\\"ASC\\",\\"DESC\\"],
});
export const SomeEnum = enumType({
  name: \\"SomeEnum\\",
  members: [\\"A\\",\\"B\\"],
});

export const UUID = scalarType({
  name: \\"UUID\\",
  asNexusMethod: \\"uuid\\",
  serialize() { /* Todo */ },
  parseValue() { /* Todo */ },
  parseLiteral() { /* Todo */ }
});"
`);
});

test("convertSDL as commonjs", () => {
  expect(convertSDL(EXAMPLE_SDL, true)).toMatchInlineSnapshot(`
"const { objectType, stringArg, arg, uuidArg, interfaceType, inputObjectType, unionType, enumType, scalarType } = require('nexus');

const Mutation = objectType({
  name: \\"Mutation\\",
  definition(t) {
    t.string(\\"someList\\", {
      list: [false],
      args: {
        items: stringArg({
          list: [false],
          required: true
        }),
      },
    })
    t.field(\\"createPost\\", {
      type: Post,
      args: {
        input: arg({
          type: CreatePostInput,
          required: true
        }),
      },
    })
    t.field(\\"registerClick\\", {
      type: Query,
      args: {
        uuid: uuidArg(),
      },
    })
  }
})
const Post = objectType({
  name: \\"Post\\",
  definition(t) {
    t.implements(Node)
    t.uuid(\\"uuid\\")
    t.field(\\"author\\", { type: User })
    t.float(\\"geo\\", { list: [true, true] })
    t.float(\\"messyGeo\\", {
      list: [true, false],
      nullable: true,
    })
  }
})
const Query = objectType({
  name: \\"Query\\",
  definition(t) {
    t.field(\\"user\\", { type: User })
    t.list.field(\\"posts\\", {
      type: Post,
      args: {
        filters: arg({
          type: PostFilters,
          required: true
        }),
      },
    })
    t.field(\\"unionField\\", { type: ExampleUnion })
  }
})
const User = objectType({
  name: \\"User\\",
  definition(t) {
    t.implements(Node)
    t.string(\\"name\\", {
      description: \\"This is a description of a name\\",
      args: {
        prefix: stringArg({ description: \\"And a description of an arg\\" }),
      },
    })
    t.string(\\"email\\")
    t.string(\\"phone\\", { nullable: true })
    t.list.field(\\"posts\\", {
      type: Post,
      args: {
        filters: arg({ type: PostFilters }),
      },
    })
    t.field(\\"outEnum\\", {
      type: SomeEnum,
      nullable: true,
    })
  }
})

const Node = interfaceType({
  name: \\"Node\\",
  description: \\"This is a description of a Node\\",
  definition(t) {
    t.id(\\"id\\")
    t.resolveType(() => null)
  }
});

const CreatePostInput = inputObjectType({
  name: \\"CreatePostInput\\",
  definition(t) {
    t.string(\\"name\\", { required: true })
    t.id(\\"author\\", { required: true })
    t.float(\\"geo\\", {
      list: [false, true],
      required: true,
    })
  }
});
const PostFilters = inputObjectType({
  name: \\"PostFilters\\",
  definition(t) {
    t.field(\\"order\\", {
      type: OrderEnum,
      required: true,
    })
    t.string(\\"search\\")
  }
});

const ExampleUnion = unionType({
  name: \\"ExampleUnion\\",
  definition(t) {
    t.members(Post, User)
  }
});

const OrderEnum = enumType({
  name: \\"OrderEnum\\",
  members: [\\"ASC\\",\\"DESC\\"],
});
const SomeEnum = enumType({
  name: \\"SomeEnum\\",
  members: [\\"A\\",\\"B\\"],
});

const UUID = scalarType({
  name: \\"UUID\\",
  asNexusMethod: \\"uuid\\",
  serialize() { /* Todo */ },
  parseValue() { /* Todo */ },
  parseLiteral() { /* Todo */ }
});

exports.Mutation = Mutation;
exports.Post = Post;
exports.Query = Query;
exports.User = User;
exports.Node = Node;
exports.CreatePostInput = CreatePostInput;
exports.PostFilters = PostFilters;
exports.ExampleUnion = ExampleUnion;
exports.OrderEnum = OrderEnum;
exports.SomeEnum = SomeEnum;
exports.UUID = UUID;"
`);
});
