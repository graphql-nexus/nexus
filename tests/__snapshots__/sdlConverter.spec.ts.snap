// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SDLConverter printEnumTypes 1`] = `
"export const OrderEnum = enumType({
  name: \\"OrderEnum\\",
  members: [\\"ASC\\",\\"DESC\\"],
});
export const SomeEnum = enumType({
  name: \\"SomeEnum\\",
  members: [\\"A\\",{\\"name\\":\\"B\\",\\"deprecation\\":\\"This is a deprecation reason for B\\",\\"value\\":\\"B\\"}],
});"
`;

exports[`SDLConverter printObjectTypes 1`] = `
"export const Mutation = objectType({
  name: \\"Mutation\\",
  definition(t) {
    t.nonNull.list.string(\\"someList\\", {
      args: {
        items: nonNull(list(stringArg())
      },
    })
    t.nonNull.field(\\"createPost\\", {
      type: Post,
      args: {
        input: arg({ type: nonNull(CreatePostInput) })
      },
    })
    t.nonNull.field(\\"registerClick\\", {
      type: Query,
      args: {
        uuid: arg({ type: UUID })
      },
    })
  }
})
export const Post = objectType({
  name: \\"Post\\",
  description: \\"This is a description of a Post\\",
  definition(t) {
    t.implements(Node)
    t.nonNull.uuid(\\"uuid\\")
    t.nonNull.field(\\"author\\", { type: User })
    t.field(\\"geo\\", { type: nonNull(list(nonNull(list(nonNull(Float))))) })
    t.field(\\"messyGeo\\", { type: list(list(nonNull(Float))) })
  }
})
export const Query = objectType({
  name: \\"Query\\",
  definition(t) {
    t.nonNull.field(\\"user\\", { type: User })
    t.field(\\"posts\\", {
      type: nonNull(list(nonNull(Post))),
      args: {
        filters: arg({ type: nonNull(PostFilters) })
      },
    })
    t.nonNull.field(\\"unionField\\", { type: ExampleUnion })
  }
})
export const User = objectType({
  name: \\"User\\",
  definition(t) {
    t.implements(Node)
    t.nonNull.string(\\"name\\", {
      description: \\"This is a description of a name\\",
      args: {
        prefix: stringArg({ description: \\"And a description of an arg\\" })
      },
    })
    t.nonNull.string(\\"email\\")
    t.string(\\"phone\\")
    t.field(\\"posts\\", {
      type: nonNull(list(nonNull(Post))),
      args: {
        filters: arg({ type: PostFilters })
      },
    })
    t.field(\\"outEnum\\", { type: SomeEnum })
  }
})"
`;

exports[`convertSDL 1`] = `
"import { objectType, stringArg, list, nonNull, arg, interfaceType, inputObjectType, unionType, enumType, scalarType } from '@nexus/schema';

export const Mutation = objectType({
  name: \\"Mutation\\",
  definition(t) {
    t.nonNull.list.string(\\"someList\\", {
      args: {
        items: nonNull(list(stringArg())
      },
    })
    t.nonNull.field(\\"createPost\\", {
      type: Post,
      args: {
        input: arg({ type: nonNull(CreatePostInput) })
      },
    })
    t.nonNull.field(\\"registerClick\\", {
      type: Query,
      args: {
        uuid: arg({ type: UUID })
      },
    })
  }
})
export const Post = objectType({
  name: \\"Post\\",
  description: \\"This is a description of a Post\\",
  definition(t) {
    t.implements(Node)
    t.nonNull.uuid(\\"uuid\\")
    t.nonNull.field(\\"author\\", { type: User })
    t.field(\\"geo\\", { type: nonNull(list(nonNull(list(nonNull(Float))))) })
    t.field(\\"messyGeo\\", { type: list(list(nonNull(Float))) })
  }
})
export const Query = objectType({
  name: \\"Query\\",
  definition(t) {
    t.nonNull.field(\\"user\\", { type: User })
    t.field(\\"posts\\", {
      type: nonNull(list(nonNull(Post))),
      args: {
        filters: arg({ type: nonNull(PostFilters) })
      },
    })
    t.nonNull.field(\\"unionField\\", { type: ExampleUnion })
  }
})
export const User = objectType({
  name: \\"User\\",
  definition(t) {
    t.implements(Node)
    t.nonNull.string(\\"name\\", {
      description: \\"This is a description of a name\\",
      args: {
        prefix: stringArg({ description: \\"And a description of an arg\\" })
      },
    })
    t.nonNull.string(\\"email\\")
    t.string(\\"phone\\")
    t.field(\\"posts\\", {
      type: nonNull(list(nonNull(Post))),
      args: {
        filters: arg({ type: PostFilters })
      },
    })
    t.field(\\"outEnum\\", { type: SomeEnum })
  }
})

export const Node = interfaceType({
  name: \\"Node\\",
  description: \\"This is a description of a Node\\",
  definition(t) {
    t.nonNull.id(\\"id\\")
  }
});

export const CreatePostInput = inputObjectType({
  name: \\"CreatePostInput\\",
  definition(t) {
    t.nonNull.string(\\"name\\")
    t.nonNull.id(\\"author\\")
    t.field(\\"geo\\", { type: nonNull(list(nonNull(list(Float)))) })
  }
});
export const PostFilters = inputObjectType({
  name: \\"PostFilters\\",
  definition(t) {
    t.nonNull.field(\\"order\\", { type: OrderEnum })
    t.string(\\"search\\", { default: \\"nexus\\" })
  }
});

export const ExampleUnion = unionType({
  name: \\"ExampleUnion\\",
  definition(t) {
    t.members(Post, User)
  }
});

export const OrderEnum = enumType({
  name: \\"OrderEnum\\",
  members: [\\"ASC\\",\\"DESC\\"],
});
export const SomeEnum = enumType({
  name: \\"SomeEnum\\",
  members: [\\"A\\",{\\"name\\":\\"B\\",\\"deprecation\\":\\"This is a deprecation reason for B\\",\\"value\\":\\"B\\"}],
});

export const UUID = scalarType({
  name: \\"UUID\\",
  asNexusMethod: \\"uuid\\",
  serialize() { /* Todo */ },
  parseValue() { /* Todo */ },
  parseLiteral() { /* Todo */ }
});"
`;

exports[`convertSDL as commonjs 1`] = `
"const { objectType, stringArg, list, nonNull, arg, interfaceType, inputObjectType, unionType, enumType, scalarType } = require('@nexus/schema');

const Mutation = objectType({
  name: \\"Mutation\\",
  definition(t) {
    t.nonNull.list.string(\\"someList\\", {
      args: {
        items: nonNull(list(stringArg())
      },
    })
    t.nonNull.field(\\"createPost\\", {
      type: Post,
      args: {
        input: arg({ type: nonNull(CreatePostInput) })
      },
    })
    t.nonNull.field(\\"registerClick\\", {
      type: Query,
      args: {
        uuid: arg({ type: UUID })
      },
    })
  }
})
const Post = objectType({
  name: \\"Post\\",
  description: \\"This is a description of a Post\\",
  definition(t) {
    t.implements(Node)
    t.nonNull.uuid(\\"uuid\\")
    t.nonNull.field(\\"author\\", { type: User })
    t.field(\\"geo\\", { type: nonNull(list(nonNull(list(nonNull(Float))))) })
    t.field(\\"messyGeo\\", { type: list(list(nonNull(Float))) })
  }
})
const Query = objectType({
  name: \\"Query\\",
  definition(t) {
    t.nonNull.field(\\"user\\", { type: User })
    t.field(\\"posts\\", {
      type: nonNull(list(nonNull(Post))),
      args: {
        filters: arg({ type: nonNull(PostFilters) })
      },
    })
    t.nonNull.field(\\"unionField\\", { type: ExampleUnion })
  }
})
const User = objectType({
  name: \\"User\\",
  definition(t) {
    t.implements(Node)
    t.nonNull.string(\\"name\\", {
      description: \\"This is a description of a name\\",
      args: {
        prefix: stringArg({ description: \\"And a description of an arg\\" })
      },
    })
    t.nonNull.string(\\"email\\")
    t.string(\\"phone\\")
    t.field(\\"posts\\", {
      type: nonNull(list(nonNull(Post))),
      args: {
        filters: arg({ type: PostFilters })
      },
    })
    t.field(\\"outEnum\\", { type: SomeEnum })
  }
})

const Node = interfaceType({
  name: \\"Node\\",
  description: \\"This is a description of a Node\\",
  definition(t) {
    t.nonNull.id(\\"id\\")
  }
});

const CreatePostInput = inputObjectType({
  name: \\"CreatePostInput\\",
  definition(t) {
    t.nonNull.string(\\"name\\")
    t.nonNull.id(\\"author\\")
    t.field(\\"geo\\", { type: nonNull(list(nonNull(list(Float)))) })
  }
});
const PostFilters = inputObjectType({
  name: \\"PostFilters\\",
  definition(t) {
    t.nonNull.field(\\"order\\", { type: OrderEnum })
    t.string(\\"search\\", { default: \\"nexus\\" })
  }
});

const ExampleUnion = unionType({
  name: \\"ExampleUnion\\",
  definition(t) {
    t.members(Post, User)
  }
});

const OrderEnum = enumType({
  name: \\"OrderEnum\\",
  members: [\\"ASC\\",\\"DESC\\"],
});
const SomeEnum = enumType({
  name: \\"SomeEnum\\",
  members: [\\"A\\",{\\"name\\":\\"B\\",\\"deprecation\\":\\"This is a deprecation reason for B\\",\\"value\\":\\"B\\"}],
});

const UUID = scalarType({
  name: \\"UUID\\",
  asNexusMethod: \\"uuid\\",
  serialize() { /* Todo */ },
  parseValue() { /* Todo */ },
  parseLiteral() { /* Todo */ }
});

exports.Mutation = Mutation;
exports.Post = Post;
exports.Query = Query;
exports.User = User;
exports.Node = Node;
exports.CreatePostInput = CreatePostInput;
exports.PostFilters = PostFilters;
exports.ExampleUnion = ExampleUnion;
exports.OrderEnum = OrderEnum;
exports.SomeEnum = SomeEnum;
exports.UUID = UUID;"
`;
